service/matching/db.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"sync"
+
+	"sync/atomic"
+
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	taskListDB struct {
+		sync.Mutex
+		domainID     string
+		taskListName string
+		taskListKind int
+		taskType     int
+		rangeID      int64
+		ackLevel     int64
+		store        persistence.TaskManager
+	}
+	taskListState struct {
+		rangeID  int64
+		ackLevel int64
+	}
+)
+
+// newTaskListDB returns an instance of an object that represents
+// persistence view of a taskList. All mutations / reads to taskLists
+// wrt persistence go through this object.
+//
+// This class will serialize writes to persistence that do condition updates. There are
+// two reasons for doing this:
+// - To work around known Cassandra issue where concurrent LWT to the same partition cause timeout errors
+// - To provide the guarantee that there is only writer who updates taskList in persistence at any given point in time
+//   This guarantee makes some of the other code simpler and there is no impact to perf because updates to tasklist are
+//   spread out and happen in background routines
+func newTaskListDB(store persistence.TaskManager, domainID string, name string, taskType int, kind int) *taskListDB {
+	return &taskListDB{
+		domainID:     domainID,
+		taskListName: name,
+		taskListKind: kind,
+		taskType:     taskType,
+		store:        store,
+	}
+}
+
+// RangeID returns the current persistence view of rangeID
+func (db *taskListDB) RangeID() int64 {
```

Comment:
Because we use rangeID to derive taskIDs. Either we expose rangeID here or we expose APIs to get current set of taskIDs - I decided to keep taskID code out of this class - because I want this class to be local view of persistence state.